src/core/components.py

```diff
@@ -805,3 +830,144 @@ def get_base_component_dict(self):
     def get_components(self):
         yield from super().get_components()
         yield from self.components
+
+
+class Container(Component):
+    """A container with optional styling and components."""
+
+    __slots__ = [
+        "components",
+        "width",
+        "style",
+    ]
+
+    def __init__(
+        self,
+        components,
+        width=12,
+        style="default",
+        **kwargs,
+    ):
+        super().__init__(**kwargs)
+        self.components = self.validate_components(components)
+        self.width = validate_size(width)
+        self.style = validate_style(style, ("transparent",))
+
+    def validate_components(self, components):
+        """Ensure no Container or ContainerRow components are added to Containers"""
+        invalid_component_types = (Container, ContainerRow)
+
+        if any(type(component) in invalid_component_types for component in components):
+            raise InvalidArguments(
+                "Container components cannot include Container Rows or other Containers."
+            )
+        return components
+
+    def get_base_component_dict(self):
+        return {
+            "type": "container",
+            "width": self.width,
+            "style": self.style,
+            "components": [
+                [component.get_flow_component_dict() for component in row]
+                for row in self.components
+            ],
+        }
+
+    def get_components(self):
+        yield from super().get_components()
+        yield from self.components
+
+    def get_validators(self):
+        yield from super().get_validators()
+
+class ContainerRow(Component):
+    """A row which contains containers."""
+
+    __slots__ = [
+        "components",
+        "height",
+    ]
+
+    def __init__(self, components, height=12, **kwargs):
+        super().__init__(**kwargs)
+        self.components = self.validate_components(components)
+        self.height = validate_size(height)
+
+    def validate_components(self, components):
+        """Ensure only Container components are added to ContainerRows"""
+        if any(not isinstance(component, Container) for component in components):
```

Comment:
Is there any technical reason for limiting this?

Reply:
For now at least we want to keep Containers and Container Rows pretty locked down. If you wanted to put components in rows you can just add them to a transparent container which will be full width by default.

Reply:
Ah, OK!  I get it now, my brain wasn't working when I read this originally.

Feel free to ignore this next bit :)

Having said that I really don't understand the need for the two components if you always have to have them together then you should probably combine them. If there is something I not see them then should probably some form of factory function which makes you a Row with a given number of Columns.